apiary-receivers-parent/apiary-receiver-common/src/main/java/com/expedia/apiary/extensions/receiver/common/messaging/JsonMetaStoreEventDeserializer.java

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (C) 2018-2019 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.expedia.apiary.extensions.receiver.common.messaging;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import com.expedia.apiary.extensions.receiver.common.error.ApiaryReceiverException;
+import com.expedia.apiary.extensions.receiver.common.event.EventType;
+import com.expedia.apiary.extensions.receiver.common.event.ListenerEvent;
+
+public class JsonMetaStoreEventDeserializer implements MetaStoreEventDeserializer {
+  private static final Logger log = LoggerFactory.getLogger(JsonMetaStoreEventDeserializer.class);
+
+  private final ObjectMapper mapper;
+
+  public JsonMetaStoreEventDeserializer(ObjectMapper mapper) {
+    this.mapper = mapper;
+  }
+
+  @Override
+  public <T extends ListenerEvent> T unmarshal(String payload) throws ApiaryReceiverException {
+    try {
+      log.debug("Marshalled event is: {}", payload);
+
+      // As we don't know the type in advance we can only deserialize the event twice:
+      // 1. Create a dummy object just to find out the type
+      T genericEvent = mapper.readerFor(HelperSerializableListenerEvent.class).readValue(payload);
+      log.debug("Unmarshal event of type: {}", genericEvent.getEventType());
+      // 2. Deserialize the actual object
+      T event = mapper.readerFor(genericEvent.getEventType().eventClass()).readValue(payload);
+      log.debug("Unmarshalled event is: {}", event);
+      return event;
+    } catch (Exception e) {
+      String message = "Unable to unmarshal event from payload";
+      throw new ApiaryReceiverException(message, e);
```

Comment:
The error here looks like it's caused when trying to deserialize the event, so shouldn't this a be a `SerDeException` that is thrown here? Also not sure why the message is declared separately on line 50, might as well just declare it directly as an argument to the exception's constructor?

Reply:
This was copied over from SY, you're right on both counts, will make the changes, thanks!